### DISCUSSION TITLE
Import multiple versions of @types/vscode into single extension

### DISCUSSION THREAD

**Discussion by vaclavHala** [microsoft/vscode-discussions#790] [Extension Development QnA] [answered]
I'm working on an extension which uses some features of recent vscode releases, but should also provide some limited functionality when installed in old vscode.

In my `package.json` I depend on the oldest `@vscode/types` supported by the extension:
```json
{
  "engines": {
    "vscode": "^1.50.0"
  },
  "devDependencies": {
    "@types/vscode": "~1.50.0"
  }
}
```

When working with code from new API I declare the types matching the new API myself, e.g.
```ts
const status = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 10) as NewStatusBarItem
// this will simply be ignored in old vscode, 
// for more complicated cases there is some conditional logic for handling the old/new API differences
status.backgroundColor = warnColor

// only present in recent versions
interface NewStatusBarItem extends vscode.StatusBarItem {
    backgroundColor?: vscode.ThemeColor
}
```

I was wondering if there is some typescript wizardry which would make it possible to have the new `vscode` namespace under some alias. I know how to download the types under alias
```json
{
  "devDependencies": {
    "@types/vscode": "~1.50.0",
    "@types/vscode-new": "npm:@types/vscode@~1.80.0"
  }
}
```
which places it in `node_modules/@types/vscode-new`, but the problem is the typescript namespace is still called `vscode` and this simple attempt
```ts
import * as vscode from 'vscode'
import * as vscodeNew from 'vscode-new'
```
does not work.

I know I could simply download the types in versions I care about, rename the namespace in the source manually and redistribute it under different name. What I'm wondering about is if there is some simple(r) way I'm not seeing

**Comment by r3m0t** [comment]
Maybe?

```
import * as vscode from 'vscode';
import type * as vscodeNew from 'vscode-new''

if (vscode.version <= '1.79') new MyBehaviour(vscode);
else new MyFancyBehaviour(vscode as typeof vscodeNew);
```

**Reply by vaclavHala** [reply]
Hmm the `import type` syntax is news to me, but the problem seems to still be the same.
I'm getting error `File '.../node_modules/@types/vscode-new/index.d.ts' is not a module.ts(2306)` on the `from 'vscode-new` part (same with both `import` and `import type`).

I suspect this is because the `'vscode'` in the declarations file is
```ts
declare module 'vscode' {
```
which in my (possibly wrong) understanding behaves differently than e.g. classes (I think your suggestion would work fine if `vscode` were a class)

**Reply by r3m0t** [reply, marked as answer]
Ah yeah, the explicit module name probably means all declarations are getting added to the same module. I guess you could use `patch-package` to change that one line?

**Reply by vaclavHala** [reply]
Nice, didn't know about patch-package and it really makes this easy.

Patching
```ts
declare module 'vscode' {
```
to
```ts
declare module 'vscode-new' {
```
makes this work
```ts
import * as vscode from 'vscode'
import * as vscodeNew from 'vscode-new'
```
as expected.

Thanks
